go/signedexchange/cbor/encoder.go

```diff
@@ -63,13 +72,126 @@ func (e *Encoder) EncodeInt(n int64) error {
+	//   represented following the rules for positive integers (major type
+	//   0).  For example, a byte string whose length is 5 would have an
+	//   initial byte of 0b010_00101 (major type 2, additional information
+	//   5 for the length), followed by 5 bytes of binary content.  A byte
+	//   string whose length is 500 would have 3 initial bytes of
+	//   0b010_11001 (major type 2, additional information 25 to indicate a
+	//   two-byte length) followed by the two bytes 0x01f4 for a length of
+	//   500, followed by 500 bytes of binary content.
+	return e.encodeBytes(TypeBytes, bs)
+}
+
+func (e *Encoder) EncodeTextString(s string) error {
+	// Major type 3:  a text string, specifically a string of Unicode
+	//   characters that is encoded as UTF-8 [RFC3629].  The format of this
+	//   type is identical to that of byte strings (major type 2), that is,
+	//   as with major type 2, the length gives the number of bytes.  This
+	//   type is provided for systems that need to interpret or display
+	//   human-readable text, and allows the differentiation between
+	//   unstructured bytes and text that has a specified repertoire and
+	//   encoding.  In contrast to formats such as JSON, the Unicode
+	//   characters in this type are never escaped.  Thus, a newline
+	//   character (U+000A) is always represented in a string as the byte
+	//   0x0a, and never as the bytes 0x5c6e (the characters "\" and "n")
+	//   or as 0x5c7530303061 (the characters "\", "u", "0", "0", "0", and
+	//   "a").
+	bs := []byte(s)
+	if !utf8.Valid(bs) {
+		return ErrInvalidUTF8
+	}
+
+	return e.encodeBytes(TypeText, bs)
+}
+
+func (e *Encoder) EncodeArrayHeader(n int) error {
+	// Major type 4:  an array of data items.  Arrays are also called lists,
+	//   sequences, or tuples.  The array's length follows the rules for
+	//   byte strings (major type 2), except that the length denotes the
+	//   number of data items, not the length in bytes that the array takes
+	//   up.  Items in an array do not need to all be of the same type.
+	//   For example, an array that contains 10 items of any type would
+	//   have an initial byte of 0b100_01010 (major type of 4, additional
+	//   information of 10 for the length) followed by the 10 remaining
+	//   items.
+	return e.encodeTypedUInt(TypeArray, uint64(n))
+}
+
+func (e *Encoder) EncodeBool(b bool) error {
+	ai := byte(0)
+	if b {
+		// True (major type 7, additional information 21)
+		ai = 21
+	} else {
+		// False (major type 7, additional information 20)
+		ai = 20
+	}
+
+	bs := []byte{TypeOther | ai}
+	if _, err := e.w.Write(bs); err != nil {
+		return err
+	}
+	return nil
+}
+
+type MapEntryEncoder struct {
+	keyBuf   bytes.Buffer
+	valueBuf bytes.Buffer
+
+	keyE   *Encoder
+	valueE *Encoder
+}
+
+func NewMapEntry() *MapEntryEncoder {
+	e := &MapEntryEncoder{}
+	e.keyE = &Encoder{&e.keyBuf}
+	e.valueE = &Encoder{&e.valueBuf}
+	return e
+}
+
+func (e *MapEntryEncoder) Bytes() []byte {
```

Comment:
KeyBytes? This is not representative of the entire map entry.

Reply:
Done